File: app/src/types/statement.ts
export type StatementCheckStatus = "DRAFT" | "UNCHECKED" | "CHECKED" | "VOID";
export type StatementInvoiceStatus = "NOT_UPLOADED" | "UPLOADED" | "REVIEWING";
export type StatementInstanceStatus =
  | "ACTIVE"
  | "COMPLETED"
  | "FAILED"
  | "CANCELLED";

export type StatementSopStepId =
  | "STMT_CREATING"
  | "STMT_SDMS_CHECK"
  | "STMT_GENERATING"
  | "STMT_PENDING_INVOICE"
  | "STMT_PENDING_REVIEW"
  | "STMT_SUBMITTED";

export type StatementPersistedStage =
  | "STMT_GENERATING"
  | "STMT_PENDING_INVOICE"
  | "STMT_PENDING_REVIEW"
  | "STMT_SUBMITTED"
  | "STMT_CANCELLED";

export interface StatementBillListItem {
  id: string;
  processInstanceId: string;
  portalAccountId: string;
  checkDate: string;
  checkAmount: string | number;
  checkStatus: StatementCheckStatus;
  invoiceStatus: StatementInvoiceStatus;
  invoiceNo?: string | null;
  invoiceAmount?: string | number | null;
  lastError?: string | null;
  stage?: StatementPersistedStage | string | null;
  instanceStatus?: StatementInstanceStatus | string | null;
  lastErrorCode?: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface StatementStageHistoryItem {
  id: string;
  fromStage?: string | null;
  toStage: string;
  actor: string;
  note?: string | null;
  createdAt: string;
}

export interface StatementSubTask {
  id: string;
  title: string;
  taskType: string;
  status: string;
  createdAt: string;
  updatedAt: string;
  lineNumber?: string | null;
}

export interface StatementBillDetail extends StatementBillListItem {
  sdmsCheckHeadId?: string | null;
  sdmsCheckNum?: string | null;
  lines?: StatementReceiptLine[];
  subTasks?: StatementSubTask[];
  stageHistory?: StatementStageHistoryItem[];
}

export interface StatementTaskResult {
  taskId: string;
  status: string;
}

export interface StatementQueryResult {
  taskId: string;
  status: string;
  runStatus?: string | null;
  rows: Record<string, unknown>[];
  errorMessage?: string | null;
}

export interface StatementGenerateResult {
  ok: boolean;
  instanceId?: string;
  taskId?: string;
  billId?: string;
  localAmount?: string;
  sdmsAmount?: string;
  sdmsCheckHeadId?: string;
  sdmsCheckNum?: string;
}

export interface StatementReceiptLine {
  receiptNo: string;
  lineNo: string;
  orderNo?: string;
  reconcileStatus?: string;
  docType?: string;
  inboundConfirmDate?: string;
  materialNumber?: string;
  itemName?: string;
  itemSpec?: string;
  receivedQty?: string;
  unitPrice?: string;
  untaxedUnitPrice?: string;
  taxRate?: string;
  untaxedAmount?: string;
  taxAmount?: string;
  taxIncludedAmount?: string;
  docDate?: string;
  billQty?: string;
  actualArrivalDate?: string;
  [key: string]: unknown;
}

File: service/app/schemas/statement.py
"""对账单 API schemas。"""

from datetime import date, datetime
from decimal import Decimal
from typing import Any

from pydantic import AliasChoices, Field

from app.schemas.common import CamelModel
from app.schemas.process import ProcessStageHistoryResponse, ProcessSubTaskResponse


class StatementQueryReceiptsRequest(CamelModel):
    portal_account_id: str = Field(
        validation_alias=AliasChoices("portal_account_id", "portalAccountId"),
        serialization_alias="portalAccountId",
    )
    date_start: str = Field(
        validation_alias=AliasChoices("date_start", "dateStart"),
        serialization_alias="dateStart",
    )
    date_end: str = Field(
        validation_alias=AliasChoices("date_end", "dateEnd"),
        serialization_alias="dateEnd",
    )


class StatementTaskResponse(CamelModel):
    task_id: str = Field(serialization_alias="taskId")
    status: str


class StatementQueryReceiptsResult(CamelModel):
    task_id: str = Field(serialization_alias="taskId")
    status: str
    run_status: str | None = Field(None, serialization_alias="runStatus")
    rows: list[dict[str, Any]] = Field(default_factory=list)
    error_message: str | None = Field(None, serialization_alias="errorMessage")


class StatementGenerateRequest(CamelModel):
    portal_account_id: str = Field(
        validation_alias=AliasChoices("portal_account_id", "portalAccountId"),
        serialization_alias="portalAccountId",
    )
    lines: list[dict[str, Any]]
    date_start: str | None = Field(
        None,
        validation_alias=AliasChoices("date_start", "dateStart"),
        serialization_alias="dateStart",
    )
    date_end: str | None = Field(
        None,
        validation_alias=AliasChoices("date_end", "dateEnd"),
        serialization_alias="dateEnd",
    )


class StatementGenerateResponse(CamelModel):
    ok: bool
    instance_id: str | None = Field(None, serialization_alias="instanceId")
    task_id: str | None = Field(None, serialization_alias="taskId")
    bill_id: str | None = Field(None, serialization_alias="billId")
    local_amount: str | None = Field(None, serialization_alias="localAmount")
    sdms_amount: str | None = Field(None, serialization_alias="sdmsAmount")
    sdms_check_head_id: str | None = Field(None, serialization_alias="sdmsCheckHeadId")
    sdms_check_num: str | None = Field(None, serialization_alias="sdmsCheckNum")


class StatementBillListItem(CamelModel):
    id: str
    process_instance_id: str = Field(serialization_alias="processInstanceId")
    portal_account_id: str = Field(serialization_alias="portalAccountId")
    check_date: date = Field(serialization_alias="checkDate")
    check_amount: Decimal = Field(serialization_alias="checkAmount")
    check_status: str = Field(serialization_alias="checkStatus")
    invoice_status: str = Field(serialization_alias="invoiceStatus")
    invoice_no: str | None = Field(None, serialization_alias="invoiceNo")
    invoice_amount: Decimal | None = Field(None, serialization_alias="invoiceAmount")
    last_error: str | None = Field(None, serialization_alias="lastError")
    stage: str | None = None
    instance_status: str | None = Field(None, serialization_alias="instanceStatus")
    last_error_code: str | None = Field(None, serialization_alias="lastErrorCode")
    created_at: datetime = Field(serialization_alias="createdAt")
    updated_at: datetime = Field(serialization_alias="updatedAt")


class StatementBillDetail(StatementBillListItem):
    sdms_check_head_id: str | None = Field(None, serialization_alias="sdmsCheckHeadId")
    sdms_check_num: str | None = Field(None, serialization_alias="sdmsCheckNum")
    lines: list[dict[str, Any]] = Field(default_factory=list)
    sub_tasks: list[ProcessSubTaskResponse] = Field(
        default_factory=list, serialization_alias="subTasks"
    )
    stage_history: list[ProcessStageHistoryResponse] = Field(
        default_factory=list, serialization_alias="stageHistory"
    )
    scanned_file_paths: list[str] = Field(
        default_factory=list, serialization_alias="scannedFilePaths"
    )


class StatementInvoicePathsRequest(CamelModel):
    file_paths: list[str] = Field(
        default_factory=list,
        validation_alias=AliasChoices("file_paths", "filePaths"),
        serialization_alias="filePaths",
    )
